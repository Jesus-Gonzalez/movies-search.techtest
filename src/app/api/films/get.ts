import axios from 'axios';

interface IObject {
  data: any;
  key: string;
}

const apiBase = 'https://swapi.dev/api';
const filmObjects = [
  'characters',
  'planets',
  'species',
  'starships',
  'vehicles',
];

export const getFilm = async (filmId: string) => {
  const film = await axios({
      url: `${apiBase}/films/${filmId}`,
      responseType: 'json'
    })
    .then(response => response.data);

  // fetch every pending object on film
  const objects: IObject[] = await Promise.all(
    filmObjects.reduce((acc: any, key: string) => (
      acc.concat(
        film[key].map((url: string) =>
          axios({ url, responseType: 'json' })
            .then(({ data }) => ({ data, key }))
        ))), []));

  objects
    .forEach(({ key }: IObject) => (
      Object.assign(film, { [key]: [] })
    ));

  objects.forEach(
    ({ data, key }) => {
        Object.assign(film, { [key]: film[key].concat(data) })
    });

  return film;
}
