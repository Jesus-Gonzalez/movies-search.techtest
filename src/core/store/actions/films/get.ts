import * as ActionTypes from 'core/store/actionTypes';

export const getFilm = (filmId: string) => ({
  type: ActionTypes.Films.GetFilm,
  filmId,
});
