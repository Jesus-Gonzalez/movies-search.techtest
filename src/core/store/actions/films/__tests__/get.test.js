import { getFilm } from '../get';

it('creates the action', () => {
  expect(getFilm('star wars'))
    .toMatchSnapshot();
});
