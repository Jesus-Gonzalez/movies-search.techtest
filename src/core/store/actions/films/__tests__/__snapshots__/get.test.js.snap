// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`creates the action 1`] = `
Object {
  "filmId": "star wars",
  "type": "GET_FILM",
}
`;
